Objects/dictobject.c

```diff
@@ -289,6 +289,47 @@ unicode_get_hash(PyObject *o)
     return ((PyASCIIObject*)o)->hash;
 }
 
+PyDict_WatchCallback
+PyDict_GetWatchCallback(void)
+{
+    PyThreadState *tstate = PyThreadState_GET();
+    return (PyDict_WatchCallback)tstate->interp->dict_watch_callback;
+}
+
+static inline int
+dict_is_watched(PyDictObject *mp)
+{
+    return mp->ma_version_tag & DICT_VERSION_WATCHED_TAG;
+}
+
+/* If we're using GCC, use __builtin_expect() to reduce overhead of
+   the watched-dict checks */
+#if defined(__GNUC__) && (__GNUC__ > 2) && defined(__OPTIMIZE__)
+#  define UNLIKELY(value) __builtin_expect((value), 0)
+#  define LIKELY(value) __builtin_expect((value), 1)
+#else
+#  define UNLIKELY(value) (value)
+#  define LIKELY(value) (value)
+#endif
+
+static inline uint64_t
+notify_dict_event(PyDict_WatchEvent event,
+                  PyDictObject *mp,
+                  PyObject *key,
+                  PyObject *value)
+{
+    // Most dicts will not be watched.
+    if (UNLIKELY(dict_is_watched(mp))) {
+        PyDict_WatchCallback callback = PyDict_GetWatchCallback();
+        // If a dict is watched, it is very likely someone installed a callback.
+        if (LIKELY(callback != NULL)) {
+            callback(event, (PyObject*)mp, key, value);
+        }
```

Comment:
for debuggability, would it be useful to raise an exception if a dict is watched on there's no callback? (at least in debug builds?)

Reply:
I think this would likely be more irritating than useful. It would mean you _must_ install your callback before any dicts can be marked as watched. But you could have a scenario where the easiest place to ensure all the dicts you want to watch are watched could occur before you've installed your callback (and this is fine for correctness, you don't actually need the notifications until you install the callback.) I don't think the debuggability benefit here outweighs the reduction in flexibility.